esmvaltool/diag_scripts/weighting/calibrate_sigmas.py

```diff
@@ -0,0 +1,282 @@
+    else:
+        denominator = 1
+
+    weights = numerator / denominator
+    weights /= weights.sum(where=~idx_nan)
+
+    return weights
+
+
+def calculate_weights_perfect_model(performance_matrix, independence,
+                                    performance_sigma, independence_sigma):
+    # TODO: this could/should probably be done in 'mask_dependent_models'
+    performance_matrix.values[np.diag_indices(
+        performance_matrix['model_ensemble'].size)] = np.nan
+    weights_matrix = xr.apply_ufunc(
+        calculate_weights_data,
+        performance_matrix,
+        independence,
+        performance_sigma,
+        independence_sigma,
+        input_core_dims=[['model_ensemble'], [], [], []],
+        output_core_dims=[['model_ensemble']],
+        vectorize=True,
+    )
+
+    return weights_matrix
+
+
+def calculate_percentiles(data, weights, percentiles=(10, 90)):
+    idx_keep = ~np.isnan(weights)
+    data = data[idx_keep]
+    weights = weights[idx_keep]
+    quantiles = np.array(percentiles) / 100
+
+    return (weighted_quantile(data, quantiles),
+            weighted_quantile(data, quantiles, weights))
+
+
+# Option 1 to calibrate sigma
+def get_confidence_score(target, percentiles, weights_matrix,
+                         performance_sigma):
+    percentiles_data, percentiles_data_weighted = xr.apply_ufunc(
+        calculate_percentiles,
+        target,
+        weights_matrix,
+        input_core_dims=[['model_ensemble'], ['model_ensemble']],
+        output_core_dims=[['percentile'], ['percentile']],
+        vectorize=True,
+        kwargs={'percentiles': percentiles})
+
+    target = target.rename({'model_ensemble': 'model_ensemble_reference'})
+
+    # inside_count_unweighted = xr.ufuncs.logical_and(
+    #     target >= percentiles_data.isel(percentile=0),
+    #     target <= percentiles_data.isel(percentile=1))
+
+    inside_count_weighted = xr.ufuncs.logical_and(
+        target >= percentiles_data_weighted.isel(percentile=0),
+        target <= percentiles_data_weighted.isel(percentile=1)).values
+
+    # inside_ratio_unweighted = inside_count_unweighted.sum() / len(inside_count_unweighted)
+    # NOTE: this is only strictly speaking true if the perfect model (nor any other model)
+    # is not removed from the ensemble. A better reference might be the 'inside_ration_unweighted'
+    inside_ratio_reference = (percentiles[1] -
+                              percentiles[0]) / 100 - 50  # DEBUG: remove - 50
+    inside_ratio_weighted = inside_count_weighted.sum() / len(
+        inside_count_weighted)
+
+    # NOTE: in the simplest case we just want the smalest sigma which is not overconfident
+    # but there might be reasons to use something slightly different
+
+    confident = inside_ratio_weighted >= inside_ratio_reference
+    # either the smallest confident sigma or the largerst sigma in total wins
+    if confident:
+        return performance_sigma
+    return 9999 - performance_sigma
+
```

Comment:
Apologies for cross posting :yum: 

Maybe this is still too simple, but I wanted to offer a thought. If we specify `confidence` as a spectrum around 0.80, where anything below 0.80 is overconfident (and lower is worse), and anything above 0.80 as 'underconfident' (diffident? ; higher is worse), then minimizing something like `abs(confidence - target_confidence)` where `target_confidence=0.80` could maybe work?